Source/Libs/GtkSharp/Widget.cs

```diff
@@ -232,13 +255,109 @@ static void ClassInit (GLib.GType gtype, Type t)
 			GLib.Marshaller.Free (native_signame);
 		}
 
+		static void InitTemplateForType (GLib.GType gtype, Type type)
+		{
+			var attr = type.GetCustomAttribute<TemplateAttribute> (true);
+			if (attr == null) return;
+
+			var data = TemplateData.Create ();
+			data.ThrowOnUnknownChild = attr.ThrowOnUnknownChild;
+			var resource_name = attr.Ui;
+			var resource_stream = type.Assembly.GetManifestResourceStream (resource_name);
+
+			if (resource_stream == null)
+				throw new Exception ("Template resource '" + resource_name + "' not found");
+
+			SetTemplateFromStream (gtype, resource_stream);
+			BindTemplateChildren (gtype, type, data.FieldBindings);
+			
+			data.SignalConnector = new SignalConnector (type);
+			data.SignalConnector.ConnectSignals (gtype);
+			Templates[type] = data;
+		}
+
+		delegate IntPtr d_gtk_widget_class_set_template(IntPtr class_ptr, IntPtr template_bytes);
+		static d_gtk_widget_class_set_template gtk_widget_class_set_template = FuncLoader.LoadFunction<d_gtk_widget_class_set_template>(FuncLoader.GetProcAddress(GLibrary.Load(Library.Gtk), "gtk_widget_class_set_template"));
+
+		static void SetTemplateFromStream (GLib.GType gtype, System.IO.Stream resource)
+		{
+			var buffer = new byte[(int)resource.Length];
+			resource.Read (buffer, 0, buffer.Length);
+			resource.Dispose ();
+
+			var bytes = new GLib.Bytes (buffer);
+			gtk_widget_class_set_template (gtype.GetClassPtr (), bytes.Handle);
+			bytes.Dispose ();
+		}
+
+		delegate IntPtr d_gtk_widget_class_bind_template_child_full(IntPtr class_ptr, IntPtr name, bool internal_child, IntPtr struct_offset);
+		static d_gtk_widget_class_bind_template_child_full gtk_widget_class_bind_template_child_full = FuncLoader.LoadFunction<d_gtk_widget_class_bind_template_child_full>(FuncLoader.GetProcAddress(GLibrary.Load(Library.Gtk), "gtk_widget_class_bind_template_child_full"));
+
+		static void BindTemplateChildren (GLib.GType gtype, Type type, Dictionary<FieldInfo, string> fields)
+		{
+			const BindingFlags flags = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.DeclaredOnly | BindingFlags.Instance;
+
+			foreach (var field in type.GetFields (flags))
+			{
+				var attr = field.GetCustomAttribute<ChildAttribute> (true);
+
+				if (attr == null)
+					continue;
+
+				var name = attr.Name ?? field.Name;
+
+				var native_name = GLib.Marshaller.StringToPtrGStrdup (name);
+				gtk_widget_class_bind_template_child_full (gtype.GetClassPtr (), native_name, attr.Internal, new IntPtr ((long)0));
+				GLib.Marshaller.Free (native_name);
+
+				fields[field] = name;
+			}
+		}
+
 		static void InitCssName (GLib.GType gtype, Type t)
 		{
 			CssNameAttribute attr = t.GetCustomAttribute<CssNameAttribute>(true);
 			if (attr != null)
 				SetCssName (gtype, attr.Name);
 		}
 
+		protected Widget () : base (IntPtr.Zero)
+		{
+			CreateNativeObject (Array.Empty<string> (), Array.Empty<GLib.Value> ());
+		}
+
```

Comment:
why this change is needed?

the autogenerated code looks like this:

```
		protected Widget() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```
it's the same, except modernizing empty array

Reply:
https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1825

Reply:
that's nice and good, but should't we do this generally in the code generator?

Reply:
i copied from develop branch:

```
#region Autogenerated code
	[GLib.TypeInitializer (typeof (Gtk.Widget), "ClassInit")]
	public partial class Widget : GLib.InitiallyUnowned, Atk.IImplementor {

		public Widget (IntPtr raw) : base(raw) {}

		protected Widget() : base(IntPtr.Zero)
		{
			CreateNativeObject (new string [0], new GLib.Value [0]);
		}
```
what i'm missing?

Reply:
btw, i'm currently reviewing generator code to fix that generally.

Reply:
You mean that generated code is same? So we don't need `<attr path="/api/namespace/object[@cname='GtkWidget']" name="disable_void_ctor">true</attr>` ?

Reply:
yes, if we change to
 `Array.Empty<string> (), Array.Empty<GLib.Value> () `
in code generator.

Reply:
I agree

Reply:
so i change code generator with a separate PR, and we can revert changes in this pr concerning `disable_void_ctor` in api

Reply:
Yes

Reply:
Good

Reply:
Reverted